src/dune_engine/dep.ml

```diff
@@ -228,34 +234,36 @@ module Set = struct
   let dir_without_files_dep dir =
     file_selector (File_selector.create ~dir Predicate.false_)
 
-  let source_tree_gen dir ~empty ~add_empty_dir ~add_paths =
+  let source_tree dir =
+    let module M = File_tree.Dir.Make_map_reduce (Memo.Build) (T) in
     let prefix_with, dir = Path.extract_build_context_dir_exn dir in
-    match File_tree.find_dir dir with
-    | None -> empty
+    File_tree.find_dir dir >>= function
+    | None -> Memo.Build.return empty
     | Some dir ->
-      File_tree.Dir.fold dir ~init:empty ~traverse:Sub_dirs.Status.Set.all
-        ~f:(fun dir acc ->
+      M.map_reduce dir ~traverse:Sub_dirs.Status.Set.all ~f:(fun dir ->
           let files = File_tree.Dir.files dir in
           let path = Path.append_source prefix_with (File_tree.Dir.path dir) in
           match String.Set.is_empty files with
-          | true -> add_empty_dir acc path
+          | true -> Memo.Build.return (singleton (dir_without_files_dep path))
           | false ->
             let paths =
               String.Set.fold files ~init:Path.Set.empty ~f:(fun fn acc ->
                   Path.Set.add acc (Path.relative path fn))
             in
-            add_paths acc paths)
-
-  let source_tree =
-    source_tree_gen ~empty ~add_paths ~add_empty_dir:(fun t dir ->
-        add t (dir_without_files_dep dir))
-
-  let source_tree_with_file_set =
-    source_tree_gen ~empty:(empty, Path.Set.empty)
-      ~add_paths:(fun (deps, paths) more_paths ->
-        (add_paths deps more_paths, Path.Set.union paths more_paths))
-      ~add_empty_dir:(fun (deps, paths) dir ->
-        (add deps (dir_without_files_dep dir), paths))
+            Memo.Build.return (add_paths empty paths))
+
+  let source_tree_with_file_set dir =
+    let+ t = source_tree dir in
+    let paths =
+      (* Here we only look at [File _]. This is ok because the above
+         [source_tree] function only produces [File _] and globs that evaluate
```

Comment:
Perhaps, it would be better if this function raises if this assumption is violated?

Reply:
You are right. I started doing that but got lazy... Let me do this.

Reply:
It's done. BTW, I remember that you mentioned something about a `Source_dir_existence` dep today, that would help here.